Initial skeleton construction

henrym2 opened the pull request:
Initial PR to setup the skeleton on the repo based on previous progress in development. 

> 

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

henrym2 committed:
Added gitignore
Ignore vendor
ignore composer.lock

henrym2 committed:
Added skeleton of the bundle structure
Current config is Active/inactive only
Reporting endpoint for Logging only
Fetch metadata start point
COOP/COEP headers start point
Dev and prod autowire services
Basic tests for Request Subscriber

henrym2 committed:
Modified Config
Changes cross_origin config to coop to make it more descriptive

henrym2 committed:
Added install and config info

rowan-m commented on composer.json:
```diff
@@ -0,0 +1,26 @@
+    "name": "ise/web-security-bundle",
+    "description": "Web security bundle for Symfony create by google OSS",
+    "type": "symfony-bundle",
+    "license": "MIT",
```
`Apache-2.0`

rowan-m commented on composer.json:
```diff
@@ -0,0 +1,26 @@
+    "authors": [
+        {
+            "name": "Matthew Henry",
+            "email": "henrym2@tcd.ie"
```
As a note for later, before this is published to Packagist we should set up a `google.com` or forwarding list for maintenance.

henrym2 replied:
For now I'll switch that out to my corp mail, and will setup something along those lines later

rowan-m commented on DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,58 @@
+        if (method_exists($treeBuilder, 'getRootNode')) {
+            $rootNode = $treeBuilder->getRootNode();
+        } else {
+            // BC for symfony/config < 4.2
```
You've specified a minimum version of `5.1` in the `composer.json`.

henrym2 replied:
Good catch, will remove this but follow Nelmio Security bundles example and allow previous versions for symfony bundles (until v 4.3).

rowan-m commented on DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,58 @@
+
+    private function getCOOP(): EnumNodeDefinition {
+        $node = new EnumNodeDefinition('coop');
+        $node->values(['none', 'cross-origin', 'same-site', 'same-origin', 'same-origin-allow-popups', 'report-only']);
```
Discussion point: With the introduction of the `SameSite=None` attribute value, I ran into the problem that libraries aggressively validated possible values and would not allow `SameSite=None`. As a result, this does lock the developer in if a new COOP value is added. I'm not wholly sure how to resolve this. Perhaps the ability to add some kind of freeform override? Or a boolean that disables checking? Or the ability to specify additional valid values?

henrym2 replied:
Not quite sure I understand this completely. I could change this to a scalar freeform field to leave it to the developer. However this would cause some issues with respect to preventing incorrect header values

rowan-m commented on EventSubscriber/RequestSubscriber.php:
```diff
@@ -0,0 +1,69 @@
+     * @param Request $req 
+     * @return void
+     */
+    private function allowRequest(Request $req) 
```
`allowRequest` seems to generic here. Isn't this just *an* example of a potential policy you would want to use for allowing/denying a request based on the `sec-fetch`*` headers?

henrym2 replied:
Absolutely too generic name wise, will be broken into a separate class possibly too. But from what I've been reading of the implementation guides that policy is the minimum of what should be included

rowan-m commented on EventSubscriber/ResponseSubscriber.php:
```diff
@@ -0,0 +1,39 @@
+    public function responseEvent(ResponseEvent $event)
+    {
+        $response = $event->getResponse();
+        if($this->active) {
```
Is this just an initial pass? It feels as if this could be encapsulated as a policy. Maybe its own class, but maybe just a configurable set of parameters in this class.

henrym2 replied:
This is just a first pass, the intention is to have the individual header values be set based on the config, as opposed to it being a binary switch. Wheither or not this includes breaking off into a separate class yet is TBD.

rowan-m commented on README.md:
```diff
@@ -8,6 +8,42 @@ including:
+
+>WIP, package not currently published. 
+
+To install the bundle on a project, add the following lines to your composer.json
```
Nice. Definitely fire up a demo server using this.

rowan-m commented on tests/RequestSubscriberTest.php:
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Ise\WebSecurityBundle\Tests;
```
Set up Travis.

henrym2 committed:
Patched based on PR comments
Added Apache license
Changed email
Updated BC for requires and require-dev

Removed BC from configuration reader

henrym2 committed:
Added Symfony/framework-bundle

henrym2 committed:
Made small scale changes
Added WIP markers for methods
Change allowRequest method name
Fixed overlooked bug with active property

henrym2 committed:
Final update pre-pull
Added phpunit to gitignore
Added travis integration
Added php.xml file
Fixed request subscriber test

rowan-m reviewed (approved):
Happy! I'll raise issues to cover future stuff, but let's get this deployed!
